### PR TITLE
Free inputs to inductor when they are no longer needed

### PR DESCRIPTION
**Pull request by eellison**
Free inputs when they are no longer needed in torchinductor. 

Follow up to https://github.com/pytorch/pytorch/pull/83137/, for more context read [this](https://github.com/pytorch/pytorch/pull/83137/#issuecomment-1211320670).

cc @anijain2305

### CODE REVIEW

File: test/inductor/test_torchinductor.py [hunk 1]
```diff
@@ -3655,6 +3656,63 @@ def fn(x, y):
...
+                    a = x + y
+                    return (a @ a,)
+
+                fn_fx = make_fx(fn)(inps[0], inps[1])
+                fn_compiled = compile_fx_inner(fn_fx, inps)
+
+                test_self = self
+                matmul_seen = False
+
+                class TestRefMode(TorchDispatchMode):
```
Comment (Chillee) [inline test/inductor/test_torchinductor.py:3688]: Lol, clever test.

**Review by Chillee — COMMENTED** [review]
Looks good to me overall, probably want to verify memory improvements before landing.

**Comment by Chillee** [comment]
@eellison I think I verified that all the references in AOTAutograd were already cleared (at least at the time of the PR), since I was able to see memory reductions with the `aot_eager` backend.

**Comment by anijain2305** [comment]
@Chillee 
image

Is this optimization already enabled for `aot_eager`? The numbers seem to suggest, that it's not. Should we try to enable it with `aot_eager` first just as a confirmation that nothing new has creeped into the memory footprint?

**Comment by eellison** [comment]
@Chillee I verified memory reduction on a small example, although there may be other issues with the larger models

**Comment by Chillee** [comment]
@eellison I've verified that this seems to clear the memory as expected using Sanket's memory profiler. On the second run, this is how the memory looks like.

image

This is what it looks like before this PR

image

There appears to still be a memory spike at the beginning of the backwards pass (before we run any code). Might be caused by some extra copying, or perhaps by Triton compilation.

image

**Review by Chillee — APPROVED** [review]

**Pull request merged**